modules/build/src/main/scala/scala/build/Inputs.scala

```diff
@@ -271,22 +288,37 @@ object Inputs {
 
     assert(validElems.nonEmpty)
 
-    val (inferredWorkspace, inferredNeedsHash, workspaceOrigin) = validElems
-      .collectFirst {
-        case d: Directory => (d.path, true, WorkspaceOrigin.SourcePaths)
+    val (inferredWorkspace, inferredNeedsHash, workspaceOrigin) = {
+      val settingsFiles = projectSettingsFiles(validElems)
+      val dirsAndFiles = validElems.collect {
+        case d: Directory  => d
+        case f: SourceFile => f
       }
-      .getOrElse {
-        validElems.head match {
-          case elem: SourceFile => (elem.path / os.up, true, WorkspaceOrigin.SourcePaths)
-          case _: Virtual =>
-            val dir = homeWorkspace(validElems, directories)
-            (dir, false, WorkspaceOrigin.HomeDir)
-          case r: ResourceDirectory =>
-            // Makes us put .scala-build in a resource directory :/
-            (r.path, true, WorkspaceOrigin.ResourcePaths)
-          case _: Directory => sys.error("Can't happen")
-        }
+
+      settingsFiles.collectFirst { s =>
+        if (settingsFiles.length > 1)
+          System.err.println(
+            s"Warning: more than one project.settings.scala file has been found. Setting ${s.base} as the project root directory for this run."
+          )
+        (s.base, true, WorkspaceOrigin.SourcePaths)
+      }.getOrElse {
+        dirsAndFiles.collectFirst {
+          case d: Directory =>
+            if (dirsAndFiles.length > 1)
+              System.err.println(
+                s"Warning: setting ${d.path} as the project root directory for this run."
+              )
+            (d.path, true, WorkspaceOrigin.SourcePaths)
+          case f: SourceFile =>
+            if (dirsAndFiles.length > 1)
+              System.err.println(
+                s"Warning: setting ${f.path / os.up} as the project root directory for this run."
+              )
+            (f.path / os.up, true, WorkspaceOrigin.SourcePaths)
+        }.getOrElse((os.pwd, true, WorkspaceOrigin.Forced))
```

Comment:
Having several nested `getOrElse` is kind of an anti-pattern too. Better use `orElse`, and a single `getOrElse` at the end, like
```scala
settingsFiles
  .collectFirst { s =>
    …
  }
  .orElse {
    dirsAndFiles.collectFirst {
      …
    }
  }
  .getOrElse {
    (os.pwd, true, WorkspaceOrigin.Forced)
  }
```